quickwit/quickwit-storage/src/object_storage/s3_compatible_storage_resolver.rs

```diff
@@ -20,22 +20,33 @@
 use std::sync::Arc;
 
 use async_trait::async_trait;
+use aws_sdk_s3::Client as S3Client;
 use quickwit_common::uri::Uri;
 use quickwit_config::{S3StorageConfig, StorageBackend};
+use tokio::sync::OnceCell;
 
+use super::s3_compatible_storage::create_s3_client;
 use crate::{
     DebouncedStorage, S3CompatibleObjectStorage, Storage, StorageFactory, StorageResolverError,
 };
 
 /// S3 compatible object storage resolver.
 pub struct S3CompatibleObjectStorageFactory {
     storage_config: S3StorageConfig,
+    // we cache the S3Client so we don't rebuild one every time we need to connect to S3.
```

Comment:
Even without this caching we didn´t create a new client for every connection as long as the `Storage` object was reused.

Reply:
it wasn't reused cross request

Reply:
This makes me wonder if the best approach wouln't be to cache the resolved storages in the [storage resolver](https://github.com/quickwit-oss/quickwit/blob/main/quickwit/quickwit-storage/src/storage_resolver.rs#L40). This would solve the problem once and for all for all storage types. But I think this solution is already good as it does solve the linked issues.

Reply:
the storages are related to a Uri (the S3Client isn't), which changes for each index. It would be interesting to be able to share the debouncer though, which isn't done here. That *could* make concurrent requests slightly more efficient.